Feature: Canonicalize `ListView`

connortsui20 opened the pull request:
Tracking Issue: https://github.com/vortex-data/vortex/issues/4699

This PR migrates the `Canonical::List` encoding from `ListArray` to `ListViewArray`.

There were a lot of small things that had to be updated, as well as tests that assumed `ListArray` was the default encoding.

# Unresolved

- [x] ~~Bug in TPC-DS (`Failed to downcast`) due to `list_contains` kernel not expecting a `ListViewArray`~~
- [X] I am not entirely sure that my refactoring of `encodings/sparse/src/canonical.rs` and `vortex-array/src/arrays/chunked/decode.rs` is correct, so I would appreciate scrutiny on that. (I cleaned it up but also filed https://github.com/vortex-data/vortex/issues/4777)
- [x] Should the `take` compute function on `ListArray` return a `ListViewArray`? We might also want to defer this question to the operator changes later. (yes)
- [X] The preferred arrow encoding for lists is still the normal `GenericListArray`, not their `GenericListViewArray`. There are some tests that rely on this, not sure if should just remove parts of those tests. (`ToArrow` is now implemented for `ListArray`
- [ ] The statpopgen benchmark slows down by 1.6-2.1x. That should be investigated, but is performance in the scope of this PR? There are a lot of compute functions that `ListView` needs to operate well (to make use of the fact that it is a view).

connortsui20 committed:
migrate `List` to use `ListViewArray` encoding

Signed-off-by: Connor Tsui <connor.tsui20@gmail.com>

connortsui20 committed:
fix list_contains for `ListView`

Signed-off-by: Connor Tsui <connor.tsui20@gmail.com>

connortsui20 commented:
A few bugs still remaining but those shouldnt impact the entire PR

robert3005 commented on vortex-array/src/arrays/chunked/decode.rs:
```diff
@@ -83,56 +83,64 @@ fn swizzle_struct_chunks(
+        // Append offsets and sizes, adjusting offsets to point into the combined array.
+        for i in 0..chunk.len() {
+            offsets.push(current_elements_offset + offsets_slice[i]);
+            sizes.push(sizes_slice[i]);
```
this doesn't have to happen in the loop, do it once for all sizes

robert3005 commented on vortex-array/src/arrays/list/compute/take.rs:
```diff
@@ -14,6 +13,8 @@ use crate::validity::Validity;
 use crate::vtable::ValidityHelper;
 use crate::{Array, ArrayRef, OffsetPType, ToCanonical, register_kernel};
 
+// TODO(connor)[ListView]: Should `take` on a `ListArray` return a `ListViewArray` instead?
```
I think so - you can return whatever is more convenient, there's no requirement that the function returns the same encoding

robert3005 commented on vortex-array/src/canonical.rs:
```diff
@@ -500,6 +500,7 @@ mod test {
     }
 
     #[test]
+    #[ignore = "TODO(connor)[ListView]: Arrow prefers `List` over `ListView`"]
```
all you need here is an `ToArrow` kernel for ListArray so it can support roundtrip

connortsui20 replied:
Actually where do I add that?

robert3005 replied:
there's already `impl ToArrowKernel for VarBinVTable` which you can use as a template

robert3005 commented on vortex-duckdb/src/exporter/list.rs:
```diff
@@ -3,86 +3,99 @@
-        // we copied all the elements on the first call to `export` so that subsequent calls need
-        // only copy the correct slice of offsets?
-        vector.list_vector_reserve(sum_of_list_lens)?;
+        // TODO(connor): This is quite bad: we have to export the entire elements array every single
```
@joseph-isaacs might have a solution here

joseph-isaacs replied:
You try what we do in the dict array, where we create a single elements vector in `new_exporter` and then `reference` that as the element child. 

I wonder if

```
let mut child = vector.list_vector_get_child()
child.reference(self.duckdb_elements)
```

Will just work, you might want to include `vector.list_vector_set_size(self.num_elements as u64)?` after

connortsui20 replied:
~~I think the set_size call might be unnecessary?~~ Edit: it was definitely necessary

robert3005 reviewed (commented):
Answered some ofthe inline comments. I think we should keep ToArrow kernel for list array and then we are good to go

connortsui20 committed:
fix sparse encoding listview

Signed-off-by: Connor Tsui <connor.tsui20@gmail.com>

connortsui20 committed:
cleanups

Signed-off-by: Connor Tsui <connor.tsui20@gmail.com>

joseph-isaacs commented on vortex-duckdb/src/exporter/list.rs:
```diff
@@ -3,86 +3,99 @@
-        vector.list_vector_set_size(sum_of_list_lens)?;
-        self.elements_exporter.export(
-            usize::try_from(start_offset)?,
-            usize::try_from(sum_of_list_lens)?,
-            &mut elements_vector,
-        )
+
+        self.elements_exporter
+            .export(0, self.num_elements, &mut elements_vector)
```
Do this after writing the elements to the child?
```suggestion
let mut elements_vector = vector.list_vector_get_child();

        self.elements_exporter
            .export(0, self.num_elements, &mut elements_vector)
            
        vector.list_vector_set_size(self.num_elements as u64)?;
```

connortsui20 committed:
make `List::take` implementation return `ListView` instead

Signed-off-by: Connor Tsui <connor.tsui20@gmail.com>

connortsui20 committed:
implement to_arrow for ListArray (again)

Signed-off-by: Connor Tsui <connor.tsui20@gmail.com>

connortsui20 committed:
make list duckdb exporter more efficient

Signed-off-by: Connor Tsui <connor.tsui20@gmail.com>

joseph-isaacs commented on vortex-duckdb/src/exporter/dict.rs:
```diff
@@ -52,6 +52,7 @@ pub(crate) fn new_exporter(
     let values_vector = match cached_vector {
         Some(vector) => vector,
         None => {
+            // TODO(connor)[ListView]: Do we not need to reserve this?
```
? reserve what?

connortsui20 commented:
@claude review

joseph-isaacs commented on vortex-array/src/compute/list_contains.rs:
```diff
@@ -268,19 +266,44 @@ fn list_contains_scalar(
+            )
+            .into_array())
+        })
     })
```
can we move this into its own function. In not a great idea to have large macro bodies

joseph-isaacs commented on vortex-duckdb/src/exporter/list.rs:
```diff
@@ -3,86 +3,105 @@
+
+    let mut duckdb_elements = Vector::with_capacity(elements.dtype().try_into()?, elements.len());
+    let elements_exporter = new_array_exporter(elements, cache)?;
+    elements_exporter.export(0, elements.len(), &mut duckdb_elements)?;
```
I know I suggested this, but it worth at least nothing the trade off here, this will export all elements, even those that are unreachable. Another approach can be seen here (https://github.com/vortex-data/vortex/issues/4767), this is the possible regression from this approach (https://github.com/vortex-data/vortex/pull/4759)

joseph-isaacs commented:
This list view exporter is really slow, do you think we should think about adding a bound value to list elements. This would record the min and max element used by the offsets in the list view. This we can be refined by slice and filter

connortsui20 committed:
fix some list regressions and address other comments

- implement list view rebuilding
- add list view exporter conversion cast
- take rebuild optimization
- hack for empty export bug

Signed-off-by: Connor Tsui <connor.tsui20@gmail.com>

connortsui20 commented:
Since the conversation is getting long and I need to rebase a bunch of stuff I'm closing this and will reopen a fresh PR soon.
